### PR TITLE
Vector RAII

### PR DESCRIPTION
**Pull request by drwells**
Spin-off from #1586.

### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

File: doc/news/changes/minor/20230504DavidWells [hunk 1]
```diff
@@ -0,0 +1,4 @@
+New: added two new classes ScopedVectorCopy and ScopedVectorDuplicate to enable
+usage of SAMRAI vectors with RAII semantics.
+<br>
+(David Wells, 2023/05/04)
```
Comment (drwells) [inline doc/news/changes/minor/20230504DavidWells:4]: We'll write a major changelog entry for the three of us once this is all done.

File: ibtk/include/ibtk/ibtk_utilities.h [hunk 2]
```diff
@@ -324,8 +326,101 @@ using FRDV = FreeRigidDOFVector;
...
+    x.resetLevels(coarsest_ln_in, finest_ln_in);
+    return;
+}
+
+/*!
+ * Utility function which asserts that the SAMRAI pointer is not null.
+ *
+ * This is useful for writing generic code in which we might want to assert that
+ * a pointer is not null in the initialization list where we cannot use the
+ * normal assertion macro. assertion.
```
Comment (boyceg) [inline ibtk/include/ibtk/ibtk_utilities.h:397]:
```suggestion
 * normal assertion macro.
```
?

File: ibtk/include/ibtk/private/ScopedVectorDuplicate-inl.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    : ScopedVectorDuplicate(checked_dereference(vector), name)
+{
+}
+
+template <typename TYPE>
+ScopedVectorDuplicate<TYPE>::ScopedVectorDuplicate(const SAMRAI::solv::SAMRAIVectorReal<NDIM, TYPE>& vector,
+                                                   const std::string& name)
+{
+    d_vector = vector.cloneVector(name);
+    d_vector->allocateVectorData();
```
Comment (boyceg) [inline ibtk/include/ibtk/private/ScopedVectorDuplicate-inl.h:47]: Do we want to allow for users to pass in a value for the `timestamp` argument to `allocateVectorData()`?

**Comment by boyceg** [comment]
Looks good. Are there more places in the library where we should switch to using this class before merging?

**Comment by boyceg** [comment]
I wonder if we should just make our own `AMRVector`, `HierarchyVector`, and/or `LevelVector` class(es) and stop using `SAMRAIVectoReal`.

File: ibtk/include/ibtk/private/ScopedVectorDuplicate-inl.h (hunk 3 above)
Comment (drwells) [reply ibtk/include/ibtk/private/ScopedVectorDuplicate-inl.h:47]: We could, but I don't think its necessary - these vectors should not participate in time interpolation.

**Comment by drwells** [comment]
I switched over one more class- that's two fewer persistent vectors. In general I think we should always copy these in `solve()` instead of setting up scratch vectors.

**Comment by boyceg** [comment]
I am not sure, but I think it would be good to avoid generating unneeded copies of the vector _data_.

File: ibtk/src/solvers/impls/FACPreconditioner.cpp [hunk 4]
```diff
@@ -94,48 +96,43 @@ FACPreconditioner::setTimeInterval(const double current_time, const double new_t
...
+    if (deallocate_after_solve) initializeSolverState(x, b);
 
     // Set the initial guess to equal zero.
-    u.setToScalar(0.0, /*interior_only*/ false);
+    x.setToScalar(0.0, /*interior_only*/ false);
+
+    // Clone the right-hand-side vector to avoid modifying it during the
+    // preconditioning operation.
+    ScopedVectorCopy<double> f(b);
+    ScopedVectorDuplicate<double> r(b);
```
Comment (drwells) [inline ibtk/src/solvers/impls/FACPreconditioner.cpp:111]: I agree that we should avoid extra data copies. What's going on here is that we set up `f` as a deep copy of `b` and `r` as a vector of the same size with all zeros. This avoids one extra data copy (we don't need to set `r = f` - that's done in the residual computation)

**Comment by drwells** [comment]
This is ready to go on my end.

**Comment by drwells** [comment]
@boyceg if you're OK with this then lets merge this and rebase the other PRs on top of it.

**Comment by boyceg** [comment]
Checking now.

**Comment by boyceg** [comment]
Looks good. Comments about "duplicate" in the scoped copy class may need some clarification. One thought is --- there are multiple different vector concepts in the code. Would it be useful to include "SAMRAIVector" in the class name? Perhaps `SAMRAIVectorScopedCopy` and `SAMRAIVectorScopedDuplicate`?

**Review by boyceg — APPROVED** [review]

**Comment by drwells** [comment]
I renamed them (it only takes a moment to update such things). I think its wise that we not have `IBTK::Vector` and `IBTK::ScopedVectorCopy` have such similar names.

File: doc/news/changes/minor/20230504DavidWells [hunk 5]
```diff
@@ -0,0 +1,4 @@
+New: added two new classes ScopedVectorCopy and ScopedVectorDuplicate to enable
```
Comment (boyceg) [inline doc/news/changes/minor/20230504DavidWells:1]:
```suggestion
New: added two new classes SAMRAIScopedVectorCopy and SAMRAIScopedVectorDuplicate to enable
```

**Comment by boyceg** [comment]
> I renamed them (it only takes a moment to update such things). I think its wise that we not have `IBTK::Vector` and `IBTK::ScopedVectorCopy` have such similar names.

Sounds good. I guess I like having `SAMRAIVector` in the name to make it clearer which vector class it is working with --- but do not have very strong feelings. (I guess it would be more clear as `SAMRAIVectorReal`, but that gets pretty verbose...)

**Comment by drwells** [comment]
We don't do anything with complex numbers so that's just needlessly wordy.

**Comment by abarret** [comment]
It's not a big deal, but this is potentially a breaking change for downstream projects that utilize the `protected` attributes in `FACPreconditioner` (like one I'm working on). It's probably worth noting this in the changelog.

**Pull request merged**